dpe/src/response.rs

```diff
@@ -161,8 +164,50 @@ pub enum DpeErrorCode {
     InvalidLocality = 0x1001,
     BadTag = 0x1002,
     MaxTcis = 0x1003,
-    PlatformError = 0x1004,
-    CryptoError = 0x1005,
-    HashError = 0x1006,
-    RandError = 0x1007,
+    Platform(PlatformError) = 0x01000000,
+    Crypto(CryptoError) = 0x02000000,
+}
+
+impl DpeErrorCode {
+    /// Get the spec-defined numeric error code. This does not include the
+    /// extended error information returned from the Platform and Crypto
+    /// implementations.
+    pub fn discriminant(&self) -> u32 {
+        // SAFETY: Because `Self` is marked `repr(u32)`, its layout is a `repr(C)` `union`
+        // between `repr(C)` structs, each of which has the `u32` discriminant as its first
+        // field, so we can read the discriminant without offsetting the pointer.
+        unsafe { *<*const _>::from(self).cast::<u32>() }
+    }
+
+    pub fn get_error_code(&self) -> u32 {
+        match self {
+            DpeErrorCode::Platform(e) => self.discriminant() | e.discriminant(),
+            DpeErrorCode::Crypto(e) => self.discriminant() | e.discriminant(),
+            _ => self.discriminant(),
+        }
+    }
+
+    /// For error variants which have extended error info returned from
+    /// underlying libraries (Platform and Crypto), return that extended error
+    /// code. For all other variants, return None.
+    ///
+    /// Reporting of detailed error information is platform-defined.
+    pub fn get_error_detail(&self) -> Option<u32> {
+        match self {
+            DpeErrorCode::Platform(e) => match e {
+                PlatformError::CertificateChainError => None,
+                PlatformError::NotImplemented => None,
+                PlatformError::IssuerNameError(code) => Some(*code),
+                PlatformError::PrintError(code) => Some(*code),
+            },
+            DpeErrorCode::Crypto(e) => match e {
+                CryptoError::AbstractionLayer(code) => Some(*code),
+                CryptoError::CryptoLibError(code) => Some(*code),
+                CryptoError::Size => None,
+                CryptoError::NotImplemented => None,
+                CryptoError::HashError(code) => Some(*code),
+            },
+            _ => None,
+        }
+    }
```

Comment:
optional: it might be worth adding a couple `From`s here as well to simplify some more of the conversions

Reply:
Good idea, that's much nicer :)